source/postprocess/boundary_densities.cc

```diff
@@ -0,0 +1,184 @@
+                  cell_at_top = true;
+                if (cell->at_boundary(f) && this->get_geometry_model().depth (cell->face(f)->center())
+                    > (this->get_geometry_model().maximal_depth() - cell->face(f)->minimum_vertex_distance()/3.))
+                  cell_at_bottom = true;
+
+                if ( cell_at_top || cell_at_bottom )
+                  {
+                    //handle surface cells
+                    fe_face_values.reinit (cell, f);
+                    fe_face_values[this->introspection().extractors.temperature]
+                    .get_function_values (this->get_solution(), in.temperature);
+                    fe_face_values[this->introspection().extractors.pressure]
+                    .get_function_values (this->get_solution(), in.pressure);
+                    fe_face_values[this->introspection().extractors.velocities]
+                    .get_function_symmetric_gradients (this->get_solution(), in.strain_rate);
+
+                    in.position = fe_face_values.get_quadrature_points();
+
+                    for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                      fe_face_values[this->introspection().extractors.compositional_fields[c]]
+                      .get_function_values(this->get_solution(),
+                                           composition_values[c]);
+                    for (unsigned int i=0; i<fe_face_values.n_quadrature_points; ++i)
+                      {
+                        for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                          in.composition[i][c] = composition_values[c][i];
+                      }
+
+                    this->get_material_model().evaluate(in, out);
+
+                    //calculate the top/bottom properties
+                    if (cell_at_top)
+                      for ( unsigned int q = 0; q < fe_face_values.n_quadrature_points; ++q)
+                        {
+                          local_top_density += out.densities[q] * fe_face_values.JxW(q);
+                          local_top_area += fe_face_values.JxW(q);
+                        }
+                    if (cell_at_bottom)
+                      for ( unsigned int q = 0; q < fe_face_values.n_quadrature_points; ++q)
+                        {
+                          local_bottom_density += out.densities[q] * fe_face_values.JxW(q);
+                          local_bottom_area += fe_face_values.JxW(q);
+                        }
+                  }
+              }
+
+      const double bottom_area = Utilities::MPI::sum( local_bottom_area, this->get_mpi_communicator() );
+      const double top_area = Utilities::MPI::sum( local_top_area, this->get_mpi_communicator() );
+
+      top_density = Utilities::MPI::sum( local_top_density, this->get_mpi_communicator() ) / top_area;
+      bottom_density = Utilities::MPI::sum( local_bottom_density, this->get_mpi_communicator() ) / bottom_area;
```

Comment:
This is not efficient -- you are communicating four times here. Please create an array of four doubles, and do the sum only once via the function that does the sum on an array.